### PR TITLE
Add support for Twitter API v2 Full-Archive Search 

### LINKED ISSUE #52: Add support for Twitter API v2 Full-Archive Search
See [conversationai#18](https://github.com/conversationai/harassment-manager/pull/18)

### PR DESCRIPTION
**Pull request by kelvinkipruto**
See [conversationai#18](https://github.com/conversationai/harassment-manager/pull/18)

Fixes #52

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -128,4 +128,4 @@
     "ts-node": "^10.7.0",
     "typescript": "~4.4.4"
   }
-}
+}
```
Comment (kilemensi) [inline package.json:131]:
```suggestion
}

```

File: src/app/social-media-item.service.ts [hunk 2]
```diff
@@ -165,23 +165,4 @@ export class SocialMediaItemService {
...
-
-  return (
-    `${date.getFullYear()}` +
-    `${(MM > 9 ? '' : '0') + MM}` +
-    `${(dd > 9 ? '' : '0') + dd}` +
-    `${(hh > 9 ? '' : '0') + hh}` +
-    `${(mm > 9 ? '' : '0') + mm}`
-  );
-}
+}
```
Comment (kilemensi) [inline src/app/social-media-item.service.ts:168]:
```suggestion
}

```

File: src/app/test_constants.ts [hunk 3]
```diff
@@ -330,4 +326,4 @@ export const TWITTER_ENTRIES: Array<ScoredItem<Tweet>> = [
       IDENTITY_ATTACK: 0.033515174,
     },
   },
-];
+];
```
Comment (kilemensi) [inline src/app/test_constants.ts:329]:
```suggestion
];

```

File: src/common-types.ts [hunk 4]
```diff
@@ -287,4 +359,4 @@ export interface ClearReportRequest {
   documentId: string;
   idToken: string;
   platform: Platform;
-}
+}
```
Comment (kilemensi) [inline src/common-types.ts:362]:
```suggestion
}

```

File: src/server/middleware/twitter.middleware.ts [hunk 5]
```diff
@@ -411,3 +596,32 @@ function getUserIdFromCredential(credential: firebase.auth.OAuthCredential) {
...
+
+// Format a millisecond-based timestamp into the YYYY-MM-DDTHH:mm:ssZ date
+// format suitable for the v2 Twitter API, as defined in:
+// https://developer.twitter.com/en/docs/twitter-api/tweets/search/api-reference/get-tweets-search-all
+function formatTimestampForV2(ms: number): string {
+  const date = new Date(ms);
+  // Remove milliseconds from the ISO string (e.g.
+  // from 2022-10-12T23:09:43.430Z to 2022-10-12T23:09:43Z.
+  return date.toISOString().substring(0, 19) + 'Z';
+}
```
Comment (kilemensi) [inline src/server/middleware/twitter.middleware.ts:627]:
```suggestion
}

```

File: src/server/server_config.template.json [hunk 6]
```diff
@@ -6,6 +6,9 @@
...
     "accountName": "",
     "username": "",
-    "password": ""
+    "password": "",
+    "appKey": "",
+    "appToken": "",
+    "bearerToken": ""
   }
-}
+}
```
Comment (kilemensi) [inline src/server/server_config.template.json:14]:
```suggestion
}

```

File: src/server/serving.ts [hunk 7]
```diff
@@ -126,7 +132,7 @@ export class Server {
 
   constructor(public config: Config) {
     if (this.config.isProduction) {
-      this.log = { write: (_s: string): void => {} };
+      this.log = { write: (_s: string): void => { } };
```
Comment (kilemensi) [inline src/server/serving.ts:135]:
```suggestion
      this.log = { write: (_s: string): void => {} };
```

File: src/server/serving.ts [hunk 8]
```diff
@@ -377,4 +383,4 @@ export class Server {
     }
     return oauthClient;
   }
-}
+}
```
Comment (kilemensi) [inline src/server/serving.ts:386]:
```suggestion
}

```

**Review by kilemensi — COMMENTED** [review]
👍🏽 

---

1. I don't see changes to `app.yml` (are we gitignoring it?),
2. Unless we've made our own customisations, can we format tweet-image.component.html similarly to how it's formatted in convesationai repo? This will cut down on number of changes.

**Comment by kelvinkipruto** [comment]
`app.yml` changes were a `node js` version upgrade which we had already done.

**Comment by kilemensi** [comment]
> `app.yml` changes were a `node js` version upgrade which we had already done.

Cool @kelvinkipruto. Can we match their version at least in this `develop` branch?

**Comment by kelvinkipruto** [comment]
@kilemensi Does this look ok?

**Comment by kilemensi** [comment]
https://github.com/CodeForAfrica/harassment-manager/pull/53#issuecomment-1404601975 @kelvinkipruto

**Comment by kelvinkipruto** [comment]
> [#53 (comment)](https://github.com/CodeForAfrica/harassment-manager/pull/53#issuecomment-1404601975) @kelvinkipruto

@kilemensi A clarification, match their `node version`?

**Comment by kilemensi** [comment]
> > [#53 (comment)](https://github.com/CodeForAfrica/harassment-manager/pull/53#issuecomment-1404601975) @kelvinkipruto
> 
> @kilemensi A clarification, match their `node version`?

Of course @kelvinkipruto ... can we match their `app.yml` in the `develop` branch?

File: app.yaml [hunk 9]
```diff
@@ -12,23 +12,20 @@
...
-runtime: nodejs14
+runtime: nodejs12
 service: default
 instance_class: F4_1G
 
 # We fetch comments 500 at a time and send requests to score them all at once,
 # so we need a reasonable number of instances for this to not be slow.
 automatic_scaling:
-  max_instances: 1
-  min_instances: 0
```
Comment (kelvinkipruto) [inline app.yaml:23]: @kilemensi Updated, but I believe we should leave these `automatic_scaling` values as we had done, or should I just match them

Comment (kilemensi) [reply app.yaml:23]: In this branch, match them exactly... Once we're ready to merge to main, we can decide which `app.yml` to keep @kelvinkipruto.

**Comment by kelvinkipruto** [comment]
Ok.
It is now fully matched.
@kilemensi

**Review by kilemensi — APPROVED** [review]
🚀

**Pull request merged**